scipy_ivp.py

```diff
@@ -0,0 +1,429 @@
+"""
+This is a modified version of the scipy.integrate.solve_ivp function from the SciPy library.
+The original function is located at `scipy/integrate/_ivp/ivp.py` in the SciPy source code.
+"""
+
+
+import inspect
+import numpy as np
+from scipy.integrate._ivp.bdf import BDF
+from scipy.integrate._ivp.radau import Radau
+from scipy.integrate._ivp.rk import RK23, RK45, DOP853
+from scipy.integrate._ivp.lsoda import LSODA
+from scipy.integrate._ivp.common import EPS, OdeSolution
+from scipy.integrate._ivp.base import OdeSolver
+import time
+import traceback
+import importlib
+from termcolor import colored
+
+
+# import scipy_jacobian
+try:
+    import scipy_jacobian
+    importlib.reload(scipy_jacobian)
+except ImportError as _error:
+    try:
+        from  Bubble_dynamics_simulation import scipy_jacobian
+        importlib.reload(scipy_jacobian)
+    except ImportError as _error:
+        print(colored(f'Error, \'scipy_ivp.py\' not found', 'red'))
+        raise _error
+    except Exception as _error:
+        print(colored(f'Error, \'scipy_ivp.py\' failed to load', 'red'))
+        raise _error
+except Exception as _error:
+    print(colored(f'Error, \'scipy_ivp.py\' failed to load', 'red'))
+    raise _error
+
+
+METHODS = {'RK23': RK23,
+           'RK45': RK45,
+           'DOP853': DOP853,
+           'Radau': Radau,
+           'BDF': BDF,
+           'LSODA': LSODA}
+
+class FunctionWrapper:
+    def __init__(self, fun, args):
+        self.raw_fun = fun
+        self.args = args
+        self.neval = 0
+        self.id = id(fun)
+
+    def __call__(self, t, x):
+        self.neval += 1
+        return self.raw_fun(t, x, *self.args)
+
+    def add_to_neval(self, n):
+        self.neval += int(n)
+
+class OdeResult:
```

Comment:
New OdeResult class. (The return of solve_ivp, usually named num_sol)
* attributes:
    * t: array, time points
     * y: array, solution values
     * nstep: int, number of successful steps
     * nfev: int, number of evaluations of the right-hand side (including nfev_jac)
     * nfev_jac: int, number of evaluations of the right-hand taken by the numerical Jacobian approximator
     * njac: int, number of evaluations of the Jacobian
     * nlu: int, number of LU decompositions
     * runtime: float, time taken to solve the ODE
     * message: str, human-readable description of the termination reason
     * success: bool, True if the solver reached the interval end or a termination event occurred
     * details: str, in case of runtime error, this field contains the traceback
     * costum: Can hold any costum attribute (currently T_max, collapse_time)
* y is transposed compared to old version (y[0] is the first step and y[:, 0] is the evolution of R)
* Has nicer string conversion and representation when you print it
  **old:**
```
message: The solver successfully reached the end of the integration interval.
  success: True
   status: 0
        t: [ 0.000e+00  6.348e-12 ...  9.999e-05  1.000e-04]
        y: [[ 3.000e-05  3.000e-05 ...  3.368e-05  3.369e-05]
            [ 0.000e+00  4.002e-07 ...  6.451e-01  6.357e-01]
            ...
            [ 0.000e+00  8.195e-97 ...  1.525e-14  1.524e-14]
            [ 0.000e+00 -3.049e-21 ...  4.475e-05  4.475e-05]]
      sol: None
 t_events: None
 y_events: None
     nfev: 63620
     njev: 1349
      nlu: 1349
```
     
  **new:**
```
message: The solver successfully reached the end of the integration interval.
  success: True
  runtime: 9.671875
        t: [0.0 ...  0.0001]
        y: [[3.00000000e-05 0.00000000e+00 2.93150000e+02 0.00000000e+00
            0.00000000e+00 3.26547855e-05 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 1.08849285e-05
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
            0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00],
           ...
           [3.36854692e-05 6.35673414e-01 2.87522097e+02 0.00000000e+00
            6.10835600e-11 2.22806725e-05 1.26490430e-15 7.03619713e-13
            9.54688020e-07 1.11841381e-14 0.00000000e+00 3.63258711e-11
            4.35181392e-16 2.21217920e-14 4.47333414e-15 4.43655975e-16
            1.69084711e-14 4.37400316e-20 3.15687847e-17 7.42682095e-06
            9.82839263e-12 1.03821624e-16 4.15638466e-13 2.00521452e-12
            2.20701855e-25 6.19380747e-20 3.59715553e-12 5.23944484e-07
            5.11168290e-18 5.38150288e-11 4.38333449e-16 9.16512188e-26
            3.98855693e-18 6.47576507e-14 1.52443289e-14 4.47540682e-05]]
    nstep: 6953
     nfev: 82413
 nfev_jac: 26866
     njac: 707
      nlu: 2318
```